dapp-core/src/main/kotlin/jp/co/soramitsu/dapp/service/ContractsRepositoryMonitor.kt

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright D3 Ledger, Inc. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package jp.co.soramitsu.dapp.service
+
+import com.d3.commons.util.createPrettySingleThreadPool
+import com.google.gson.JsonParser
+import io.reactivex.Observable
+import io.reactivex.schedulers.Schedulers
+import io.reactivex.subjects.PublishSubject
+import iroha.protocol.BlockOuterClass
+import jp.co.soramitsu.dapp.config.DAPP_NAME
+import jp.co.soramitsu.dapp.listener.ReliableIrohaChainListener
+import jp.co.soramitsu.iroha.java.QueryAPI
+import mu.KLogging
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.stereotype.Component
+
+@Component
+class ContractsRepositoryMonitor(
+    @Autowired
+    private val chainListener: ReliableIrohaChainListener,
+    @Autowired
+    private val queryAPI: QueryAPI,
+    @Autowired
+    private val dAppAccountId: String,
+    @Autowired
+    private val repositoryAccountId: String,
+    @Autowired
+    private val repositorySetterId: String
+) {
+    private val jsonParser = JsonParser()
+    private val newContractsSubject: PublishSubject<Pair<String, String>> = PublishSubject.create()
+    private val disabledContractsSubject: PublishSubject<String> = PublishSubject.create()
+    private val scheduler = Schedulers.from(createPrettySingleThreadPool(DAPP_NAME, "observable-contracts"))
+
+    fun initObservable() {
+        logger.info("Subscribed to contracts status updates")
+        chainListener.processIrohaBlocks(this::processBlock, scheduler)
+    }
+
+    private fun processBlock(block: BlockOuterClass.Block) {
```

Comment:
Please, process block only once, and select logic handler inside processing instead of separate block processing for different observers.